internal/shared/shared.go

```diff
@@ -0,0 +1,249 @@
+// allows us to obtain the non-garbled export data of all dependencies, useful
+// for type checking of the packages as we obfuscate them.
+//
+// Note that we obtain this data once in saveListedPackages, store it into a
+// temporary file via gob encoding, and then reuse that file in each of the
+// garble processes that wrap a package compilation.
+type listedPackages map[string]*ListedPackage
+
+// ListedPackage contains information useful for obfuscating a package
+type ListedPackage struct {
+	ImportPath string
+	Export     string
+	Deps       []string
+	ImportMap  map[string]string
+
+	// TODO(mvdan): reuse this field once TOOLEXEC_IMPORTPATH is used
+	private bool
+}
+
+// SetListedPackages get information about the current package
+// and all of its dependencies
+func SetListedPackages(flags, patterns []string) error {
+	args := []string{"list", "-json", "-deps", "-export"}
+	args = append(args, flags...)
+	args = append(args, patterns...)
+	cmd := exec.Command("go", args...)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("go list error: %v", err)
+	}
+	dec := json.NewDecoder(stdout)
+	data.ListedPackages = make(listedPackages)
+	for dec.More() {
+		var pkg ListedPackage
+		if err := dec.Decode(&pkg); err != nil {
+			return err
+		}
+		data.ListedPackages[pkg.ImportPath] = &pkg
+	}
+
+	if err := cmd.Wait(); err != nil {
+		return fmt.Errorf("go list error: %v: %s", err, stderr.Bytes())
+	}
+
+	anyPrivate := false
+	for path, pkg := range data.ListedPackages {
+		if isPrivate(path) {
+			pkg.private = true
+			anyPrivate = true
+		}
+	}
+
+	if !anyPrivate {
+		return fmt.Errorf("GOPRIVATE=%q does not match any packages to be built", os.Getenv("GOPRIVATE"))
+	}
+	for path, pkg := range data.ListedPackages {
+		if pkg.private {
+			continue
+		}
+		for _, depPath := range pkg.Deps {
+			if data.ListedPackages[depPath].private {
+				return fmt.Errorf("public package %q can't depend on obfuscated package %q (matched via GOPRIVATE=%q)",
+					path, depPath, os.Getenv("GOPRIVATE"))
+			}
+		}
+	}
+
+	return nil
+
+}
+
+// ListPackage get the listedPackage information for a certain package
+func ListPackage(fromPath, path string) (*ListedPackage, error) {
+	pkg, ok := data.ListedPackages[path]
+	if !ok {
+		if fromPkg, ok := data.ListedPackages[fromPath]; ok {
+			if path2 := fromPkg.ImportMap[path]; path2 != "" {
+				return ListPackage(fromPath, path2)
+			}
+		}
+		return nil, fmt.Errorf("path not found in listed packages: %s", path)
+	}
+	return pkg, nil
+}
+
+const neverPrivate = "runtime,internal/cpu,internal/bytealg,unsafe"
+
+// isPrivate checks if GOPRIVATE matches path.
+//
+// To allow using garble without GOPRIVATE for standalone main packages, it will
+// default to not matching standard library packages.
+//
+// This was copied from garble/main.go
+func isPrivate(path string) bool {
```

Comment:
I don't agree with copy-pasting this func, especially because we'll likely have to update it over time.

You could just move the "anyPrivate" check back into the main package, because it really has nothing to do with sharing data between processes. It's just a sanity check done by the main process.